src/providers/VisibleTournamentsProvider.js

```diff
@@ -0,0 +1,68 @@
+import {useMemo, useRef, useState} from 'react';
+
+import VisibleTournamentsContext from '../contexts/VisibleTournamentsContext';
+import VisibleTournamentsTrackingContext from '../contexts/VisibleTournamentsTrackingContext';
+
+import useDebounce from '../use/useDebounce';
+
+const VISIBILITY_CHANGE_DEBOUNCE_WAIT_MS = 250;
+
+const VisibleTournamentsProvider = ({ children }) => {
+    const [visibleTournaments, setVisibleTournaments] = useState(new Set());
+
+    const watchedElementToTournamentRef = useRef(new Map());
+
+    const [setVisibleTournamentsDebounced] = useDebounce(setVisibleTournaments, VISIBILITY_CHANGE_DEBOUNCE_WAIT_MS);
+
+    const intersectionObserverRef = useRef(
+        new IntersectionObserver((entries) => {
+            setVisibleTournamentsDebounced((visibleTournaments) => {
+                const currentlyVisibleTournaments = new Set(visibleTournaments);
+
+                entries.forEach((entry) => {
```

Comment:
How often is this being invoked? On every rerender? When the map is moved? It's not entirely clear to me. This could be my own ignorance talking, which is fine. If you think someone more experienced in react might be wondering the same thing, consider adding a comment somewhere explaining what is invoking this and when/how.

Reply:
Added a small comment. Overall I think this should be pretty straightforward if you are familiar with the intersection observer API.